build.gradle

```diff
@@ -152,12 +153,79 @@ tasks.whenTaskAdded {task ->
     }
 }
 
+def splitTestConfig = [
```

Comment:
This looks highly manual in a way we could miss tests.  How hard would it be to split on test class if a number of parallel runners was provided?

Reply:
This approach will never miss tests, as tests which are not matched by these patterns will fall in to the bucket that is executed by the general `test` task.

> How hard would it be to split on test class if a number of parallel runners was provided?

Not sure if I understand. Can you elaborate?

Generally, it should be noted that this is a quick-win approach as asked for by @davidlago . As described in https://github.com/opensearch-project/security/issues/2798#issuecomment-1568539046 we also would like to restructure the tests themselves to consume fewer resources. However, that won't be something we can achieve within a couple of weeks.

Reply:
This is only manual for test which we decided to split. Any new test classess will be containted within default "test" job.

Reply:
We did some profiling before and found that the execution time is not really equally distributed over the classes and packages. There are some very heavy classes and some more lightweight ones. The partitioning in this PR has been chosen to balance the execution time.

Reply:
In line with what Peter mentioned, is there a way we can perform this in a load balanced way? I know you mentioned identifying the tests manually, but I am unclear on what will happen if new tests are added. Will they just go into the default catch all? If we can find a way to balance the runners based on load we could optimize performance even when new tests are added. 

One possibility would be to track the execution time of the previous run and direct the splitting script to move tests from one runner to another based on the last execution. This would lead to improved performance for small changes where no significant tests were added to the repo since we could use the roughly optimized version of the test loads from the previous run.

Reply:
> One possibility would be to track... [Good idea]

tracking / analyzing / acting seems like next step(s) that we might want to revisit if this is a problem

Reply:
I just want to double-check: Is this now a discussion about next steps or are these requests for this PR? :-)

Reply:
> In line with what Peter mentioned, is there a way we can perform this in a load balanced way?

No.

> Will they just go into the default catch all?

Yes.

> If we can find a way to balance the runners based on load we could optimize performance even when new tests are added.

You are right BUT Github Actions does not have a interface to do that. We would have to:
1. Keep old test runs somewhere besides Github.
2. Analyze them (by what? what metric should be used?)
3. Somehow inject the new code into the gradle? (how?)
4. Who will maintain this Rube-Goldberg machine of configs and parsers?
We have issues with long-running tasks right now, adding this complicated system will take weeks if not months. Creating manual configuration is faster and easier and solves problem by a little bit. Remember that this is not a final solution. This is only a patch for currently-somewhat-broken configuration. Core issue is within tests itself, not with how they are run.

Reply:
Nothing in this thread is a requirement for merging from me. Even if we have 'imbalanced' or 'may-require-maintenance-in-time' solution its still a great improvement to accelerate time to merge

Reply:
@scrawfor99 What do you think about resolving this thread?  I don't think this thread has much actionable that we'd like to see from @pawel-gudel-eliatra

Would you want to open up an follow up issue on the test rebalancing / monitoring?

Reply:
hi @peternied, definitely. We can follow up in a separate PR. I do not want to slow @pawel-gudel-eliatra's great work with this.